### PR TITLE
Add uint32_t i32 field to Value union

### PR DESCRIPTION
**Pull request by chfast**
Problems this introduces
- `stack.top() = 0` only zeros the `.i32` now.
- `FizzyValue{1}` only inits the `.i32` part. This is excessively used in C API tests. And cannot be disabled. In C++20 you can do `FizzyValue{.i64 = 1}`.

### CODE REVIEW

**Comment by axic** [comment]
What was the blocker on this? It would make the Rust API a tiny bit nicer too.

**Comment by chfast** [comment]
> What was the blocker on this? It would make the Rust API a tiny bit nicer too.

Unit tests expectations update.

File: lib/fizzy/value.hpp
Comment (axic) [inline lib/fizzy/value.hpp:57]: Do we want to keep this?

Comment (chfast) [reply lib/fizzy/value.hpp:57]: No, as this is invalid now.

**Comment by axic** [comment]
This is now blocked on explicitly typing number literals in the tests.

File: lib/fizzy/value.hpp
Comment (axic) [inline lib/fizzy/value.hpp]: Added these as "quick fix" to see which ones the compiler couldn't correctly match :)

File: CMakeLists.txt
Comment (axic) [inline CMakeLists.txt:46]: This is just stray from my local needs. (Need to update xcode, the current one still doesn't work without this.).

File: test/utils/asserts.hpp
Comment (axic) [inline test/utils/asserts.hpp:98]: Eventually we should also remove this, if we merge this PR.

File: lib/fizzy/stack.hpp
Comment (gumb0) [inline lib/fizzy/stack.hpp:118]: Hm, I'm confused by union initialization again. Does default initialization like this set the active member of the union?

Comment (chfast) [reply lib/fizzy/stack.hpp:118]: I'm confused too. This was only found by GCC coverage build.

The `0` is definitely wrong because it calls the special constructor and inits `.i32` only.

But the default constructor should be wrong too because it should only init the first field and `.i32` is the first one. This works however for current compilers. I wander if MSan can help here. Would be good if you can check the spec yourself.

One solution to solve this is to provide own implementation of the default constructor which inits `.i64` (or define `i64 = 0;`). But with that the `Value` looses some properties like "is trivially constructable".

Second solution is to move `.i64` to the first position. This guarantees the `Value{}` to be fully zero-initialized. Same for `FizzyValue` in C API.

But I don't see any sanity solution to init locals with zero without knowing the type.

Comment (chfast) [reply lib/fizzy/stack.hpp:118]: I don't believe I'm doing it again to myself, but:
1. The `Value{}` is value-initialization.
2. if T is a class type with a default constructor that is neither user-provided nor deleted (that is, it may be a class with an implicitly-defined or defaulted default constructor), the object is zero-initialized.
3. Zero-initialization: If T is a union type, the first non-static named data member is zero-initialized and all padding is initialized to zero bits.

So I think the `Value{}` is correct here as it zeros all object's bytes.

File: lib/fizzy/execute.cpp
Comment (gumb0) [inline lib/fizzy/execute.cpp:1202]: I don't insist it's much better, but this way it would look more like you're just returning 64-bit value, without assuming internal representation of stack item
```suggestion
                stack.top() = uint64_t{0};  // FIXME: Review all cases like this.
```

File: test/spectests/spectests.cpp
Comment (gumb0) [inline test/spectests/spectests.cpp:488]: I think `read_integer_value` function has to be modified, too, because now it produces always 64-bit `Value`

File: test/utils/asserts.hpp
Comment (gumb0) [inline test/utils/asserts.hpp:103]: I think this should also include `int32_t`?

File: test/utils/asserts.hpp
Comment (gumb0) [inline test/utils/asserts.hpp:105]: comment is outdated

File: test/utils/fizzy_c_engine.cpp
Comment (gumb0) [inline test/utils/fizzy_c_engine.cpp:113]: You can add now
```
    assert(args.size() == func_type.inputs_size);
```

File: test/unittests/value_test.cpp
Comment (gumb0) [inline test/unittests/value_test.cpp:16]: Test for constructor from i32 is now needed.

File: test/unittests/capi_test.cpp
Comment (gumb0) [inline test/unittests/capi_test.cpp:596]:
```suggestion
    EXPECT_EQ(fizzy_execute(instance, 0, nullptr, 0), CResult(0x22bbaa_u32));
```

File: lib/fizzy/stack.hpp
Comment (gumb0) [reply lib/fizzy/stack.hpp:118]: It seems to depend on the interpretation of what is "padding" for a union - whether unused bits of unactive members are unions. I see some people having doubts about this here: https://stackoverflow.com/questions/11812555/how-to-zero-initialize-an-union#comment15704643_11812606 I have not found a definite answer to this yet.

However, I had a concern about something else (perhaps not directly related to this PR): we zero-initialize `Value`s of locals here, which might or might not choose the active member of the union (I have not found the answer to this either). Then function may access them right away with `local.get`, then some other operation may access some member of it - can be any type member. Then this may or may not be UB.

We can add a test for this, but as I remember sanitizers didn't catch union type punning.

And if we're required to set the right type, then there's no way around needing the types of locals when initializing.

Comment (gumb0) [reply lib/fizzy/stack.hpp:118]: > We can add a test for this, but as I remember sanitizers didn't catch union type punning.

Well, the tests you added here (`execute.local_init`) already check this actually. Because checking the result of a function reads some specific member of `Value` union.

Comment (chfast) [reply lib/fizzy/stack.hpp:118]: I did some experiments and it looks like GCC and MSVC inits the whole object while Clang only the first element: https://godbolt.org/z/cWscY8.

Comment (gumb0) [reply lib/fizzy/stack.hpp:118]: This may be a good reason to make `i64` the first member. Are there any advantages for it to be the second other than aesthetical?

**Comment by axic** [comment]
Needs rebase after similar PRs (#655, #659) are merged.

File: lib/fizzy/execute.cpp
Comment (axic) [inline lib/fizzy/execute.cpp:1432]: Do you want to add tests first before this PR?

Comment (axic) [reply lib/fizzy/execute.cpp:1432]: Done in #680 ?

Comment (chfast) [reply lib/fizzy/execute.cpp:1432]: Yes, although this PR will extend the tests anyway by polluting the Value with some garbage (not possible before).

File: test/unittests/api_test.cpp
Comment (gumb0) [inline test/unittests/api_test.cpp:47]:
```suggestion
    const ExecutionResult result = Value{1234_u32};
```

File: lib/fizzy/execute.cpp
Comment (axic) [reply lib/fizzy/execute.cpp:1202]: I'd agree this is cleaner, since the `rem()` case below is defined as `T rem(T a, T b)`, so both would be assigning a `int64_t` to `top()`.

Comment (axic) [reply lib/fizzy/execute.cpp:1202]: In that sense also I do not get the FIXME, having `top() = uint64_t{0}` seems to be the correct approach.

File: lib/fizzy/execute.cpp
Comment (axic) [inline lib/fizzy/execute.cpp:1438]: Why not?

```suggestion
            stack.top() = uint64_t{stack.top().i32};
```

File: test/unittests/api_test.cpp
Comment (axic) [inline test/unittests/api_test.cpp:65]: @gumb0 How about the `value{1234}` above?

File: test/unittests/capi_test.cpp
Comment (axic) [inline test/unittests/capi_test.cpp:384]:
```suggestion
    EXPECT_THAT(fizzy_execute(instance, 1, nullptr, 0), CResult(43_u64));
```

File: test/unittests/execute_numeric_test.cpp
Comment (axic) [inline test/unittests/execute_numeric_test.cpp:485]: Hmm, what about a case where `i64` has top bit set, but `i32` does not have top bit set. That should clear the top bits.

File: test/utils/asserts.hpp
Comment (axic) [reply test/utils/asserts.hpp:146]: @chfast do you want to remove it in a separate commit, or a separate PR? I can do the removal too.

Comment (axic) [reply test/utils/asserts.hpp:146]: Well, actually shouldn't all the test code be using `as<uint32_t>()`? We can apply that parallel to this PR.

File: test/unittests/capi_test.cpp
Comment (chfast) [reply test/unittests/capi_test.cpp:632]: Well the `CResult()` and `Result()` are pretty useless now as they try to compare all 64 bits without knowing the type.

File: lib/fizzy/execute.cpp
Comment (chfast) [reply lib/fizzy/execute.cpp]: Done.

File: test/unittests/execute_numeric_test.cpp
Comment (chfast) [reply test/unittests/execute_numeric_test.cpp:487]: Done.

File: test/unittests/value_test.cpp
Comment (chfast) [reply test/unittests/value_test.cpp:30]: Done.

File: test/utils/asserts.hpp
Comment (chfast) [reply test/utils/asserts.hpp:105]: As changed in #687, only `_u32` and `_u64` are accepted for `CResult()`.

File: test/utils/fizzy_c_engine.cpp
Comment (chfast) [reply test/utils/fizzy_c_engine.cpp:113]: Done.

File: test/unittests/api_test.cpp
Comment (gumb0) [reply test/unittests/api_test.cpp]: The way `Value` constuctor behaves with `int` should be checked by `Value` unit tests.

**Comment by chfast** [comment]
I don't see any significant speed difference except for noise. The GCC LTO build is much noisier.
```
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     +0.0095         +0.0095            80            81            80            81
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    +0.0085         +0.0085          1207          1217          1207          1217
fizzy/execute/ecpairing/onepoint_mean                             +0.0072         +0.0072        404734        407646        404736        407648
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   +0.0201         +0.0201           101           103           101           103
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  +0.0131         +0.0131          1489          1508          1489          1508
fizzy/execute/memset/256_bytes_mean                               -0.0167         -0.0167             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             -0.0209         -0.0209          1539          1507          1539          1507
fizzy/execute/mul256_opt0/input1_mean                             +0.0347         +0.0347            29            30            29            30
fizzy/execute/ramanujan_pi/33_runs_mean                           -0.0274         -0.0274           114           111           114           111
fizzy/execute/sha1/512_bytes_rounds_1_mean                        +0.0303         +0.0303            88            91            88            91
fizzy/execute/sha1/512_bytes_rounds_16_mean                       +0.0334         +0.0334          1225          1266          1225          1266
fizzy/execute/sha256/512_bytes_rounds_1_mean                      +0.0075         +0.0075            80            81            80            81
fizzy/execute/sha256/512_bytes_rounds_16_mean                     +0.0087         +0.0087          1101          1111          1101          1111
fizzy/execute/taylor_pi/pi_1000000_runs_mean                      -0.0011         -0.0011         40619         40573         40620         40573
fizzy/execute/micro/eli_interpreter/exec105_mean                  -0.0150         -0.0150             5             4             5             4
fizzy/execute/micro/factorial/20_mean                             +0.0207         +0.0206             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0175         +0.0175          5048          5137          5048          5137
fizzy/execute/micro/host_adler32/1_mean                           +0.0211         +0.0211             0             0             0             0
fizzy/execute/micro/host_adler32/1000_mean                        -0.0042         -0.0042            31            31            31            31
fizzy/execute/micro/icall_hash/1000_steps_mean                    +0.0033         +0.0033            69            69            69            69
fizzy/execute/micro/spinner/1_mean                                -0.0603         -0.0603             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             -0.0134         -0.0134             9             9             9             9
```

File: test/utils/asserts.hpp
Comment (gumb0) [reply test/utils/asserts.hpp:105]: Hm, writing a test expecting a negative value will be a bit awkward then, istn't it?

Comment (chfast) [reply test/utils/asserts.hpp:105]: `-1_u32`.

File: test/unittests/value_test.cpp
Comment (gumb0) [inline test/unittests/value_test.cpp:38]: This is the same as in the test above, did you mean:
```suggestion
    constexpr auto max_int = std::numeric_limits<int>::max();
```

File: test/unittests/value_test.cpp
Comment (gumb0) [inline test/unittests/value_test.cpp:47]: You could add here a negative value, as in new tests above

File: test/utils/asserts.hpp
Comment (gumb0) [reply test/utils/asserts.hpp:105]: But isn't it equivalent to `-uint32_t{1}`, so signed value?

Comment (gumb0) [reply test/utils/asserts.hpp:105]: I'm confused about why `Result` and `CResult` are different in this regard.
Is `Result` allowing signed values because otherwise there're many tests to change?

Comment (chfast) [reply test/utils/asserts.hpp:105]: > But isn't it equivalent to `-uint32_t{1}`, so signed value?

It is, but it stays `uint32_t` type.



> I'm confused about why `Result` and `CResult` are different in this regard.
> Is `Result` allowing signed values because otherwise there're many tests to change?

The `Result()` requires type information. I restricted values to `CResult()` more because the `EXPECT_THAT(Value{0x100000000_u64}, CResult(0_u32))` is a match.

Comment (gumb0) [reply test/utils/asserts.hpp:105]: > `EXPECT_THAT(Value{0x100000000_u64}, CResult(0_u32))` is a match.

How does restricting to only unsigned values help this?

Comment (chfast) [reply test/utils/asserts.hpp:105]: You can't put `CResult(0)` so you need to thing if it should be `_u32` or `_u64`. This also affects existing tests.

Comment (gumb0) [reply test/utils/asserts.hpp:105]: I see. And what is the disadvantage of making `Result()` accept only unsigned expected types?

Comment (chfast) [reply test/utils/asserts.hpp:105]: It may be good idea, but this requires 176 additional test changes. So I went with adding type checking instead. Can be done later if desired.

We can also require type in `CResult` or even merge these two. But I believe there is additional research needed here. E.g. we should try combining matchers: `Result: Not(Traps) && HasResult && ResultValueIs(x)`.

File: test/utils/asserts.hpp
Comment (chfast) [reply test/utils/asserts.hpp:156]: > Well, actually shouldn't all the test code be using `as<uint32_t>()`? We can apply that parallel to this PR.

The `Result()` was checking this to some distinct. But I went with additional type checking to solved it for good.

File: test/utils/asserts.hpp
Comment (gumb0) [reply test/utils/asserts.hpp:105]: I see, that's what I guessed in the beginning, too many tests to change to make `Result` accept only unsigned,

**Review by gumb0 — APPROVED** [review]
Looks fine to me :shipit:

File: test/utils/asserts.hpp
Comment (chfast) [reply test/utils/asserts.hpp:156]: > @chfast do you want to remove it in a separate commit, or a separate PR? I can do the removal too.

In https://github.com/wasmx/fizzy/pull/691.

File: test/unittests/api_test.cpp
Comment (axic) [inline test/unittests/api_test.cpp:50]: Notpick, we do not actually need `_u32` here, right?

File: test/utils/fizzy_c_engine.cpp
Comment (axic) [inline test/utils/fizzy_c_engine.cpp:120]: Hmm both of these (fizzy_c_engine and fizzy_engine) don't validate that f32/f64 is, but `bench` ensures that. Anyway perhaps we should have an assert somewhere, can be done in a new PR.

**Review by axic — APPROVED** [review]

File: test/unittests/api_test.cpp
Comment (chfast) [reply test/unittests/api_test.cpp:50]: No. But using `_u32` gives higher confidence (i.e. someone has thought what the result type is expected). Still gray area though.

File: test/utils/fizzy_c_engine.cpp
Comment (chfast) [reply test/utils/fizzy_c_engine.cpp:122]: Added asserts.

**Review by axic — APPROVED** [review]

File: include/fizzy/fizzy.h
Comment (axic) [inline include/fizzy/fizzy.h:25]: Nit: extra new line between these two, but not the others :)

**Review by gumb0 — APPROVED** [review]

**Pull request merged**
